nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,772 @@
+        mut size: u16,
+    ) -> usb_device::Result<&'static mut [u8]> {
+        if self.unalloc_buffers.len() < usize::from(size) {
+            Err(UsbError::EndpointMemoryOverflow)
+        } else {
+            if ep_type == EndpointType::Bulk || ep_type == EndpointType::Interrupt {
+                // datasheet: buffer must be 4-byte aligned and its size must be a multiple of 4
+                let rem = self.unalloc_buffers.as_mut_ptr() as usize % 4;
+                if rem != 0 {
+                    let (_padding, remaining) =
+                        mem::replace(&mut self.unalloc_buffers, &mut []).split_at_mut(4 - rem);
+                    self.unalloc_buffers = remaining;
+                }
+
+                let rem = size % 4;
+                if rem != 0 {
+                    size = size + 4 - rem;
+                }
+            }
+            assert!(size <= 64);
+            let (alloc, remaining) =
+                mem::replace(&mut self.unalloc_buffers, &mut []).split_at_mut(size.into());
+            self.unalloc_buffers = remaining;
+            Ok(alloc)
+        }
+    }
+
+    fn is_used(&self, ep: EndpointAddress) -> bool {
+        if ep.index() == 8 {
+            // ISO
+            let flag = if ep.is_in() {
+                self.iso_in_used
+            } else {
+                self.iso_out_used
+            };
+            return flag;
+        }
+
+        if ep.is_in() {
+            self.used_in & (1 << ep.index()) != 0
+        } else {
+            self.used_out & (1 << ep.index()) != 0
+        }
+    }
+
+    fn read_control_setup(&self, regs: &USBD, buf: &mut [u8], ep0_state: &mut EP0State) -> usb_device::Result<usize> {
+        const SETUP_LEN: usize = 8;
+
+        if buf.len() < SETUP_LEN {
+            return Err(UsbError::BufferOverflow);
+        }
+
+        // This is unfortunate: Reassemble the nicely split-up setup packet back into bytes, only
+        // for the usb-device code to copy the bytes back out into structured data.
+        // The bytes are split across registers, leaving a 3-Byte gap between them, so we couldn't
+        // even memcpy all of them at once. Weird peripheral.
+        buf[0] = regs.bmrequesttype.read().bits() as u8;
+        buf[1] = regs.brequest.read().brequest().bits();
+        buf[2] = regs.wvaluel.read().wvaluel().bits();
+        buf[3] = regs.wvalueh.read().wvalueh().bits();
+        buf[4] = regs.windexl.read().windexl().bits();
+        buf[5] = regs.windexh.read().windexh().bits();
+        buf[6] = regs.wlengthl.read().wlengthl().bits();
+        buf[7] = regs.wlengthh.read().wlengthh().bits();
+
+        ep0_state.direction = match regs.bmrequesttype.read().direction().is_host_to_device() {
+            false => UsbDirection::In,
+            true => UsbDirection::Out,
+        };
+
+        ep0_state.is_set_address = (buf[0] == 0x00) && (buf[1] == 0x05);
+
+        if regs.bmrequesttype.read().direction().is_host_to_device() {
```

Comment:
nit: this is a bit shorter and feels easier to understand after the assignment just above

```suggestion
        if ep0_state.direction == UsbDirection::Out  {
```